// nativescript-ui-chart/visualization/views/chart-series.js
import * as seriesCommonModule from './chart-series.common';
// NOTE: This is a dummy file used to fix Typescript errors while developing.
// The NativeScript build overwrites it with one of the corresponding .ios.ts or .android.ts
// files from the same folder.
export class PieSeries extends seriesCommonModule.PieSeries {
}
export class DonutSeries extends seriesCommonModule.DonutSeries {
}
////////////////////////////////////////////////////////////////////////
// BarSeries
////////////////////////////////////////////////////////////////////////
export class BarSeries extends seriesCommonModule.BarSeries {
}
////////////////////////////////////////////////////////////////////////
// RangeBarSeries
////////////////////////////////////////////////////////////////////////
export class RangeBarSeries extends seriesCommonModule.RangeBarSeries {
}
////////////////////////////////////////////////////////////////////////
// LineSeries
////////////////////////////////////////////////////////////////////////
export class LineSeries extends seriesCommonModule.CategoricalSeries {
}
export class SplineSeries extends LineSeries {
}
export class AreaSeries extends LineSeries {
}
export class SplineAreaSeries extends LineSeries {
}
export class BubbleSeries extends seriesCommonModule.BubbleSeries {
}
export class ScatterSeries extends seriesCommonModule.ScatterSeries {
}
export class ScatterBubbleSeries extends seriesCommonModule.ScatterBubbleSeries {
}
export class OhlcSeries extends seriesCommonModule.OhlcSeries {
}
export class CandlestickSeries extends OhlcSeries {
}
