// nativescript-auth0/android/provider/authorizeResult.js
var Activity = android.app.Activity;
var Log = android.util.Log;
export class AuthorizeResult {
    constructor() {
        if (arguments.length >= 2) {
            this.requestCode = arguments[0];
            this.resultCode = arguments[1];
            this.intent = arguments[2];
        }
        else {
            this.requestCode = AuthorizeResult.MISSING_REQUEST_CODE;
            this.resultCode = Activity.RESULT_OK;
            this.intent = arguments[0];
        }
    }
    /**
     * Checks if the received data is valid and can be parsed.
     *
     * @param expectedRequestCode the request code this activity is expecting to receive
     * @return whether if the received uri data can be parsed or not.
     */
    isValid(expectedRequestCode) {
        const uri = this.intent != null ? this.intent.getData() : null;
        if (uri == null) {
            Log.d(AuthorizeResult.TAG, "Result is invalid: Received Intent's Uri is null.");
            return false;
        }
        if (this.requestCode === AuthorizeResult.MISSING_REQUEST_CODE) {
            return true;
        }
        const fromRequest = this.getRequestCode() === expectedRequestCode;
        if (!fromRequest) {
            Log.d(AuthorizeResult.TAG, java.lang.String.format("Result is invalid: Received Request Code doesn't match the expected one. Was %d but expected %d", [this.getRequestCode(), expectedRequestCode]));
        }
        return fromRequest && this.resultCode === Activity.RESULT_OK;
    }
    getIntent() {
        return this.intent;
    }
    getRequestCode() {
        return this.requestCode;
    }
    getResultCode() {
        return this.resultCode;
    }
}
AuthorizeResult.TAG = 'AuthorizeResult';
AuthorizeResult.MISSING_REQUEST_CODE = -100;
