// nativescript-auth0/android/provider/customTabsOptions.js
var Parcelable = android.os.Parcelable;
var CustomTabsIntent = androidx.browser.customtabs.CustomTabsIntent;
var ContextCompat = androidx.core.content.ContextCompat;
var CustomTabsOptions = /** @class */ (function (_super) {
    __extends(CustomTabsOptions, _super);
    function CustomTabsOptions() {
        var _this = _super.call(this) || this;
        return global.__native(_this);
    }
    CustomTabsOptions_1 = CustomTabsOptions;
    CustomTabsOptions.create = function (showTitle, toolbarColor) {
        if (showTitle === void 0) { showTitle = false; }
        if (toolbarColor === void 0) { toolbarColor = 0; }
        var options = new CustomTabsOptions_1();
        options.showTitle = showTitle;
        options.toolbarColor = toolbarColor;
        return options;
    };
    CustomTabsOptions.prototype.toIntent = function (context, session) {
        var builder = new CustomTabsIntent.Builder(session)
            .setShowTitle(this.showTitle);
        if (this.toolbarColor > 0) {
            // Resource exists
            builder.setToolbarColor(ContextCompat.getColor(context, this.toolbarColor));
        }
        return builder.build().intent;
    };
    CustomTabsOptions.prototype.describeContents = function () {
        return 0;
    };
    CustomTabsOptions.prototype.writeToParcel = function (dest, flags) {
        dest.writeByte(this.showTitle ? 0x01 : 0x00);
        dest.writeInt(this.toolbarColor);
    };
    var CustomTabsOptions_1;
    CustomTabsOptions.CREATOR = new Parcelable.Creator({
        createFromParcel: function (parcel) {
            var showTitle = parcel.readByte() !== 0x00;
            var toolbarColor = parcel.readInt();
            return CustomTabsOptions_1.create(showTitle, toolbarColor);
        },
        newArray: function (size) {
            return Array.create(CustomTabsOptions_1, size);
        }
    });
    CustomTabsOptions = CustomTabsOptions_1 = __decorate([
        Interfaces([Parcelable])
    ], CustomTabsOptions);
    return CustomTabsOptions;
}(java.lang.Object));
export { CustomTabsOptions };
