// nativescript-auth0/android/authentication/request/signUpRequest.js
/**
 * Represent a request that creates a user in a Auth0 Database connection and then logs in.
 */
export class SignUpRequest {
    constructor(signUpRequest, authenticationRequest) {
        this.signUpRequest = signUpRequest;
        this.authenticationRequest = authenticationRequest;
    }
    /**
     * Add additional parameters sent when creating a user.
     *
     * @param parameters sent with the request and must be non-null
     * @see ParameterBuilder
     * @return itself
     */
    addSignUpParameters(parameters) {
        this.signUpRequest.addParameters(parameters);
        return this;
    }
    /**
     * Add additional parameters sent when logging the user in
     *
     * @param parameters sent with the request and must be non-null
     * @return itself
     * @see ParameterBuilder
     */
    addAuthenticationParameters(parameters) {
        this.authenticationRequest.addAuthenticationParameters(parameters);
        return this;
    }
    setScope(scope) {
        this.authenticationRequest.setScope(scope);
        return this;
    }
    setAudience(audience) {
        this.authenticationRequest.setAudience(audience);
        return this;
    }
    setAccessToken(accessToken) {
        this.authenticationRequest.setAccessToken(accessToken);
        return this;
    }
    setGrantType(grantType) {
        this.authenticationRequest.setGrantType(grantType);
        return this;
    }
    setRealm(realm) {
        this.signUpRequest.setConnection(realm);
        this.authenticationRequest.setRealm(realm);
        return this;
    }
    /**
     * Starts to execute create user request and then logs the user in.
     *
     * @param callback called on either success or failure.
     */
    start(callback) {
        this.signUpRequest.start({
            onSuccess: (user) => {
                this.authenticationRequest.start(callback);
            },
            onFailure: (error) => {
                callback.onFailure(error);
            }
        });
    }
}
