// nativescript-ui-chart/visualization/parts/chart-label-style.common.js
import { ViewBase, Property, Color } from "@nativescript/core";
export class PointLabelStyle extends ViewBase {
}
PointLabelStyle.strokeWidthProperty = new Property({
    name: "strokeWidth",
    defaultValue: undefined,
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.strokeColorProperty = new Property({
    name: "strokeColor",
    defaultValue: undefined,
    equalityComparer: Color.equals,
    valueConverter: (v) => new Color(v),
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.fillColorProperty = new Property({
    name: "fillColor",
    defaultValue: undefined,
    equalityComparer: Color.equals,
    valueConverter: (v) => new Color(v),
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.textColorProperty = new Property({
    name: "textColor",
    defaultValue: undefined,
    equalityComparer: Color.equals,
    valueConverter: (v) => new Color(v),
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.textSizeProperty = new Property({
    name: "textSize",
    defaultValue: 10,
    valueConverter: parseFloat,
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.marginProperty = new Property({
    name: "margin",
    defaultValue: undefined,
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.textFormatProperty = new Property({
    name: "textFormat",
    defaultValue: undefined,
    valueChanged: (target, oldValue, newValue) => {
    },
});
PointLabelStyle.fontNameProperty = new Property({
    name: "fontName",
    defaultValue: undefined,
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.fontStyleProperty = new Property({
    name: "fontStyle",
    defaultValue: undefined,
    valueChanged: (target, oldValue, newValue) => { },
});
PointLabelStyle.strokeWidthProperty.register(PointLabelStyle);
PointLabelStyle.strokeColorProperty.register(PointLabelStyle);
PointLabelStyle.fillColorProperty.register(PointLabelStyle);
PointLabelStyle.textColorProperty.register(PointLabelStyle);
PointLabelStyle.textSizeProperty.register(PointLabelStyle);
PointLabelStyle.marginProperty.register(PointLabelStyle);
PointLabelStyle.textFormatProperty.register(PointLabelStyle);
PointLabelStyle.fontNameProperty.register(PointLabelStyle);
PointLabelStyle.fontStyleProperty.register(PointLabelStyle);
