// nativescript-auth0/android/util/telemetry.js
var Base64 = android.util.Base64;
export class Telemetry {
    constructor(name, version, libraryVersion = undefined) {
        this.name = name;
        this.version = version;
        this.libraryVersion = libraryVersion;
    }
    getName() {
        return this.name;
    }
    getVersion() {
        return this.version;
    }
    getLibraryVersion() {
        return this.libraryVersion;
    }
    getValue() {
        const values = {};
        if (this.name != null) {
            values[Telemetry.NAME_KEY] = this.name;
        }
        if (this.version != null) {
            values[Telemetry.VERSION_KEY] = this.version;
        }
        if (this.libraryVersion != null) {
            values[Telemetry.LIB_VERSION_KEY] = this.libraryVersion;
        }
        if (Object.keys(values).length === 0) {
            return null;
        }
        const javaString = new java.lang.String(JSON.stringify(values));
        return Base64.encodeToString(javaString.getBytes('UTF-8'), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
    }
}
Telemetry.HEADER_NAME = "Auth0-Client";
Telemetry.NAME_KEY = "name";
Telemetry.VERSION_KEY = "version";
Telemetry.LIB_VERSION_KEY = "lib_version";
